### ISSUE TITLE
Deprecations on subclass reverse() cause breakage

### ISSUE THREAD

**Issue by jodygarnett** [locationtech/jts#560]
A couple of issues and one compatibility error have been noticed due to the implementation of reverseInternal() in #513

* Geometry.reverseInternal() - this method requires some javadoc guidance for implementors
* GeometryCollection.reverse(), etc - subclass `reverse` methods are deprecated, which is causing warnings and build failures in downstream applications
* LinearRing has inconsistent visibility on reverseInternal()

```java
  /** @deprecated */
  public Geometry reverse()
  {
    return super.reverse();
  }

  public Geometry reverseInternal() {
    CoordinateSequence seq = points.copy();
    CoordinateSequences.reverse(seq);
    return getFactory().createLinearRing(seq);
  }
```
The API change above was focused on a minimal change to JTS codebase, and had an unexpected consequence for downstream code.

The above inconsistencies are causing some warnings in client code like this (where MultiLineString reverse is deprecated method):

```
MultiLineString reversed = (MultiLineString) linestring.reverse();
```

And can be addressed by casting to Geometry (which is undeprecated):

```
MultiLineString reversed = (MultiLineString)((Geometry)linestring).reverse();
```

# Option 1: Remove subclass implementations of reverse()

Remove downstream code of the form shown in GeometryCollection:
```
  /**
   * Creates a {@link GeometryCollection} with
   * every component reversed.
   * The order of the components in the collection are not reversed.
   *
   * @return a {@link GeometryCollection} in the reverse order
   * @deprecated
   */
 public Geometry reverse() {
    return super.reverse();
  }
```

This ensures all references to reverse() use the one provided by Geometry:

```
MultiLineString reversed = (MultiLineString) linestring.reverse();
```

This option is the minimal change that addresses the deprecation issue.

# Option 2: Un-deprecate subclass implementations of reverse() and type narrow (Recommended)

Undeprecate and type narrow:
```
 public MultiLineString reverse() {
    return (MultiLineString ) super.reverse();
  }
```

Downstream code has less work to do:

```
MultiLineString reversed = linestring.reverse();
```

This option is recommended, it does involve some work but results in the easiest experience for library users.

# Option 3: Fun with generics (not recommended)

```
public abstract class <G extends Geometry> Geometry<G>
    implements Cloneable, Comparable, Serializable {

  public G reverse() {

    Geometry res = reverseInternal();
    if (this.envelope != null)
      res.envelope = this.envelope.copy();
    res.setSRID(getSRID());

    return res;
  }
  protected abstract G reverseInternal();
}
```

This option is not recommended: While this would technically produce the same results as *Un-deprecate subclass implementations of reverse() and type narrow* the extra bother of generics does not earn its keep. Client could would get a lot of warnings asking to be changed to `Geometry<?>` or similar.

**Comment by dr-jts** [comment]
> The warnings look like:
> 
> ```
> MultiLineString reversed = (MultiLineString) linestring.reverse();
> ```

Did you mean to paste the text of a compiler warning here?

**Comment by jodygarnett** [comment]
I was just sharing code that produces a warning

**Comment by dr-jts** [comment]
Another option to fix this:

* Add `abstract Geometry.reverse()`
* Move the common field-copying logic into a method `protected Geometry.copyFields(Geometry source, boolean copyUserData)`
* Create typed subclass `reverse()` methods
  * move logic from `reverseInternal`
  * call `revCopy.copyFields(this, false)` (user data is not copied)

This provides typed methods and factors out common logic, and makes the call chain a bit simpler.

This pattern could also be used to refactor the implementation for `copy`, to provide a typed copy method.

**Comment by dr-jts** [comment]
Fixed by #582
